File: analysis/pca_logratio_2d.py
#!/usr/bin/env python3

import argparse
import json
import subprocess
import sys
from collections import OrderedDict
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from matplotlib import rcParams
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import StandardScaler

rcParams["font.family"] = "MS Gothic"

EPS = 1e-12
DEFAULT_TSV_PATTERN = "*.tsv"


def parse_args(argv: Optional[List[str]] = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Run 2D PCA clustering on logRatio vectors derived from substitution TSV files."
    )
    parser.add_argument(
        "--tsv-dir",
        required=True,
        help=(
            "Directory containing substitution TSV files (e.g. outputs of generate_tsv_files.sh). "
            f"Files should follow the pattern '{DEFAULT_TSV_PATTERN}' unless --tsv-pattern is provided."
        ),
    )
    parser.add_argument(
        "--output-dir",
        help="Directory where PCA outputs will be written. Defaults to <tsv-dir>/pca_logratio_output.",
    )
    parser.add_argument(
        "--tsv-pattern",
        default=DEFAULT_TSV_PATTERN,
        help=f"Glob pattern to select TSV files within the TSV directory (default: {DEFAULT_TSV_PATTERN}).",
    )
    parser.add_argument(
        "--max-clusters",
        type=int,
        default=10,
        help="Maximum number of clusters to evaluate when selecting the optimal k.",
    )
    parser.add_argument(
        "--random-state",
        type=int,
        default=42,
        help="Random seed used for clustering.",
    )
    if argv is None:
        argv = sys.argv[1:]
    if not argv:
        parser.print_help(sys.stderr)
        print(
            "\nExample:\n  python scripts/analysis/pca_logratio_2d.py "
            "--tsv-dir /path/to/tsv_dir --output-dir /path/to/output_dir",
            file=sys.stderr,
        )
        raise SystemExit(1)
    return parser.parse_args(argv)


def collect_tsv_files(tsv_dir: Path, pattern: str) -> List[Path]:
    files = sorted(tsv_dir.glob(pattern))
    if not files:
        raise FileNotFoundError(f"No TSV files matching pattern '{pattern}' were found in {tsv_dir}")
    return files


def parse_filename_parts(tsv_path: Path) -> Tuple[str, str, str]:
    name_parts = tsv_path.stem.split("_")
    if len(name_parts) < 3:
        raise ValueError(f"Unexpected filename format for TSV file: {tsv_path.name}")
    accession = name_parts[0]
    short_name = name_parts[1]
    date = name_parts[2]
    return accession, short_name, date


def run_datasets_cli(accession: str) -> dict:
    try:
        proc = subprocess.run(
            ["datasets", "summary", "genome", "accession", accession],
            check=True,
            capture_output=True,
            text=True,
        )
    except (subprocess.CalledProcessError, FileNotFoundError) as exc:
        raise RuntimeError(f"Failed to fetch metadata for accession {accession}: {exc}") from exc

    output = proc.stdout.strip()
    if not output:
        raise RuntimeError(f"Empty metadata response for accession {accession}")

    try:
        return json.loads(output)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"Invalid JSON received for accession {accession}: {exc}") from exc


def ensure_metadata(tsv_path: Path, accession: str) -> dict:
    metadata_path = tsv_path.with_name(f"{accession}.json")
    if metadata_path.exists():
        with metadata_path.open("r", encoding="utf-8") as handle:
            return json.load(handle)

    metadata = run_datasets_cli(accession)
    with metadata_path.open("w", encoding="utf-8") as handle:
        json.dump(metadata, handle, ensure_ascii=False, indent=2)
    return metadata


def extract_genus(metadata: dict, short_name: str, accession: str) -> Tuple[str, str]:
    reports = metadata.get("reports", [])
    if reports:
        organism = reports[0].get("organism", {})
        organism_name = organism.get("organism_name")
        if organism_name:
            genus = organism_name.split()[0]
            label = f"{organism_name.replace(' ', '_')} ({short_name})"
            return genus, label

    fallback_label = f"{short_name} ({accession})"
    return "Unknown", fallback_label


def compute_logratio_vector(df: pd.DataFrame) -> np.ndarray:
    required_columns = {"mutType", "mutNum", "totalRootNum"}
    if missing := required_columns - set(df.columns):
        raise ValueError(f"TSV file is missing required columns: {', '.join(sorted(missing))}")

    df = df.copy()
    totals = df["totalRootNum"].to_numpy(dtype=float)
    mut_counts = df["mutNum"].to_numpy(dtype=float)
    with np.errstate(divide="ignore", invalid="ignore"):
        observed = np.divide(mut_counts, totals, out=np.zeros_like(mut_counts, dtype=float), where=totals != 0)
    mean_observed = np.mean(observed) if observed.size else 0.0
    # Use max(mean_observed, EPS) to avoid division by zero or producing extremely large ratios
    # when mean_observed is zero or extremely close to zero. Ensures numerical stability.
    adjusted_mean = max(mean_observed, EPS)
    ratios = (observed + EPS) / adjusted_mean
    with np.errstate(divide="ignore"):
        logratio = np.log2(ratios)
    return logratio


def load_tsv_logratio(tsv_path: Path, reference_order: Optional[List[str]]) -> Tuple[np.ndarray, List[str]]:
    df = pd.read_csv(tsv_path, sep="\t")
    if reference_order is None:
        # Sort by mutType to ensure consistent ordering of mutation types,
        # which is crucial for reproducible PCA/projection and for aligning
        # feature vectors across multiple samples.
        df_sorted = df.sort_values("mutType").reset_index(drop=True)
        order = df_sorted["mutType"].tolist()
        vector = compute_logratio_vector(df_sorted)
        return vector, order

    df_indexed = df.set_index("mutType")
    try:
        df_aligned = df_indexed.loc[reference_order].reset_index()
    except KeyError as exc:
        raise ValueError(f"TSV file {tsv_path.name} does not contain expected mutType values: {exc}") from exc
    vector = compute_logratio_vector(df_aligned)
    return vector, reference_order


def determine_clusters(data: np.ndarray, max_clusters: int, random_state: int) -> Tuple[int, List[float], List[float]]:
    sample_count = data.shape[0]
    if sample_count < 2:
        raise ValueError("At least two samples are required for clustering.")

    adjusted_max = min(max_clusters, sample_count)
    if adjusted_max < 2:
        adjusted_max = 2

    inertia_values: List[float] = []
    silhouette_scores: List[float] = []

    for k in range(2, adjusted_max + 1):
        model = KMeans(n_clusters=k, random_state=random_state, n_init=10)
        labels = model.fit_predict(data)
        inertia_values.append(model.inertia_)
        silhouette_scores.append(silhouette_score(data, labels))

    best_k = silhouette_scores.index(max(silhouette_scores)) + 2
    return best_k, inertia_values, silhouette_scores


def plot_metrics(output_dir: Path, inertia: List[float], silhouette: List[float]) -> None:
    ks = list(range(2, 2 + len(inertia)))
    fig, axes = plt.subplots(1, 2, figsize=(12, 5))

    axes[0].plot(ks, inertia, "bo-")
    axes[0].set_xlabel("Number of Clusters")
    axes[0].set_ylabel("Inertia")
    axes[0].set_title("Elbow Method")
    axes[0].grid(True)

    axes[1].plot(ks, silhouette, "ro-")
    axes[1].set_xlabel("Number of Clusters")
    axes[1].set_ylabel("Silhouette Score")
    axes[1].set_title("Silhouette Scores")
    axes[1].grid(True)

    plt.tight_layout()
    fig.savefig(output_dir / "clustering_metrics.png", dpi=300, bbox_inches="tight")
    plt.close(fig)


def build_color_map(genus_labels: List[str]) -> Dict[str, Tuple[float, float, float, float]]:
    unique_genus = sorted(OrderedDict.fromkeys(genus_labels))
    cmap = plt.get_cmap("tab20")
    color_map: Dict[str, Tuple[float, float, float, float]] = {}
    for idx, genus in enumerate(unique_genus):
        color_map[genus] = cmap(idx % cmap.N)
    return color_map


def plot_pca_scatter(
    output_dir: Path,
    pca_data: np.ndarray,
    labels: List[str],
    genus_labels: List[str],
    cluster_labels: np.ndarray,
) -> None:
    color_map = build_color_map(genus_labels)
    colors = [color_map[genus] for genus in genus_labels]

    fig, ax = plt.subplots(figsize=(14, 12))
    ax.scatter(pca_data[:, 0], pca_data[:, 1], c=colors, alpha=0.8, s=80, edgecolors="k")

    for idx, text in enumerate(labels):
        ax.annotate(text, (pca_data[idx, 0], pca_data[idx, 1]), xytext=(5, 5), textcoords="offset points", fontsize=8)

    ax.set_xlabel("PC1")
    ax.set_ylabel("PC2")
    ax.set_title(f"K-means Clustering (k={len(set(cluster_labels))})")
    ax.grid(True)

    handles = []
    for genus, color in color_map.items():
        handles.append(plt.Line2D([0], [0], marker="o", color="w", markerfacecolor=color, markersize=10, label=genus))
    ax.legend(handles=handles, title="Genus", loc="best")

    fig.savefig(output_dir / "pca_2d_clusters.png", dpi=300, bbox_inches="tight")
    plt.close(fig)


def save_results(
    output_dir: Path,
    files: List[Path],
    pca_data: np.ndarray,
    cluster_labels: np.ndarray,
    metadata_records: List[dict],
) -> None:
    rows: List[dict] = []
    for idx, file_path in enumerate(files):
        record = {
            "file": file_path.name,
            "accession": metadata_records[idx]["accession"],
            "short_name": metadata_records[idx]["short_name"],
            "date": metadata_records[idx]["date"],
            "genus": metadata_records[idx]["genus"],
            "cluster": int(cluster_labels[idx]),
            "PC1": float(pca_data[idx, 0]),
            "PC2": float(pca_data[idx, 1]),
        }
        rows.append(record)

    df = pd.DataFrame(rows)
    df.to_csv(output_dir / "clustering_results.csv", index=False)


def main() -> None:
    args = parse_args()

    tsv_dir = Path(args.tsv_dir).resolve()
    if not tsv_dir.is_dir():
        print(f"Error: TSV directory does not exist: {tsv_dir}", file=sys.stderr)
        sys.exit(1)

    output_dir = Path(args.output_dir).resolve() if args.output_dir else tsv_dir / "pca_logratio_output"
    output_dir.mkdir(parents=True, exist_ok=True)

    tsv_files = collect_tsv_files(tsv_dir, args.tsv_pattern)

    vectors: List[np.ndarray] = []
    metadata_records: List[dict] = []
    labels: List[str] = []
    genus_labels: List[str] = []
    reference_order: Optional[List[str]] = None

    for tsv_file in tsv_files:
        accession, short_name, date = parse_filename_parts(tsv_file)
        metadata_json = ensure_metadata(tsv_file, accession)
        genus, label = extract_genus(metadata_json, short_name, accession)

        vector, reference_order = load_tsv_logratio(tsv_file, reference_order)

        vectors.append(vector)
        labels.append(label)
        genus_labels.append(genus)
        metadata_records.append(
            {
                "accession": accession,
                "short_name": short_name,
                "date": date,
                "genus": genus,
            }
        )

    data_matrix = np.vstack(vectors)
    scaler = StandardScaler()
    scaled_data = scaler.fit_transform(data_matrix)

    pca = PCA(n_components=2)
    pca_data = pca.fit_transform(scaled_data)

    best_k, inertia_values, silhouette_scores = determine_clusters(pca_data, args.max_clusters, args.random_state)
    plot_metrics(output_dir, inertia_values, silhouette_scores)

    final_model = KMeans(n_clusters=best_k, random_state=args.random_state, n_init=10)
    cluster_labels = final_model.fit_predict(pca_data)

    plot_pca_scatter(output_dir, pca_data, labels, genus_labels, cluster_labels)
    save_results(output_dir, tsv_files, pca_data, cluster_labels, metadata_records)

    print("PCA clustering completed successfully.")
    print(f"Processed files: {len(tsv_files)}")
    print(f"Optimal clusters (k): {best_k}")
    print(f"Results saved to: {output_dir}")


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"Error: {exc}", file=sys.stderr)
        sys.exit(1)

File: analysis/R/pca_logratio_2d.R
#!/usr/bin/env Rscript

# PCA clustering pipeline that mirrors scripts/analysis/pca_logratio_2d.py.
# Usage example:
#   Rscript scripts/analysis/R/pca_logratio_2d.R \
#     --tsv-dir /path/to/tsv_dir --output-dir /path/to/output_dir
# Required packages: argparse, jsonlite, tidyverse, cluster, scales, factoextra, ggpubr.
# The script calls `datasets summary genome accession "<accession>"` when metadata JSON is missing.

suppressPackageStartupMessages({
  library(argparse)
  library(jsonlite)
  library(readr)
  library(dplyr)
  library(purrr)
  library(tidyr)
  library(stringr)
  library(ggplot2)
  library(tibble)
  library(cluster)
  library(scales)
  library(factoextra)
  library(ggpubr)
  library(rlang)
})

EPS <- 1e-12

# Suppress R CMD check notes for dplyr/ggplot2 pronouns
utils::globalVariables(c("mutNum", "totalRootNum", "mutType", "k", "PC1", "PC2", "genus", "species"))

parse_args <- function() {
  parser <- ArgumentParser(description = "Run 2D PCA clustering on substitution TSV files (logRatio vectors).")
  parser$add_argument("--tsv-dir", required = TRUE, help = "Directory containing substitution TSV files.")
  parser$add_argument("--output-dir", required = FALSE, help = "Directory to write PCA outputs (default: <tsv-dir>/pca_logratio_output).")
  parser$add_argument("--tsv-pattern", default = "*.tsv", help = "Glob pattern for TSV files (default: *.tsv).")
  parser$add_argument("--max-clusters", type = "integer", default = 10L, help = "Maximum number of clusters to evaluate.")
  parser$add_argument("--random-state", type = "integer", default = 42L, help = "Random seed for clustering.")
  args <- parser$parse_args()
  args
}

collect_tsv_files <- function(tsv_dir, pattern) {
  files <- list.files(path = tsv_dir, pattern = glob2rx(pattern), full.names = TRUE, recursive = FALSE)
  if (length(files) == 0) {
    stop(sprintf("No TSV files matching pattern '%s' were found in %s", pattern, tsv_dir))
  }
  sort(files)
}

parse_filename_parts <- function(file_path) {
  name_parts <- strsplit(sub("\\.tsv$", "", basename(file_path)), "_")[[1]]
  if (length(name_parts) < 4) {
    stop(sprintf("Unexpected filename format for TSV file: %s", basename(file_path)))
  }
  list(
    accession = paste(name_parts[1:2], collapse = "_"),
    short_name = name_parts[3],
    date = name_parts[4]
  )
}

run_datasets_cli <- function(accession) {
  command <- sprintf('datasets summary genome accession "%s"', accession)
  result <- system(command, intern = TRUE, ignore.stderr = TRUE)
  if (length(result) == 0) {
    stop(sprintf("Empty metadata response for accession %s", accession))
  }
  fromJSON(paste(result, collapse = "\n"))
}

ensure_metadata <- function(tsv_path, accession) {
  metadata_path <- file.path(dirname(tsv_path), paste0(accession, ".json"))
  if (file.exists(metadata_path)) {
    return(fromJSON(metadata_path, simplifyVector = FALSE))
  }
  metadata <- run_datasets_cli(accession)
  write_json(metadata, metadata_path, auto_unbox = TRUE, pretty = TRUE)
  metadata
}

extract_genus_label <- function(metadata, short_name, accession) {
  reports <- metadata$reports
  if (!is.null(reports) && length(reports) > 0) {
    organism <- reports[[1]]$organism
    if (!is.null(organism$organism_name)) {
      species_name <- organism$organism_name
      genus <- strsplit(species_name, " +")[[1]][1]
      return(list(genus = genus, species = species_name))
    }
  }
  list(genus = "Unknown", species = accession)
}

compute_feature_vectors <- function(df) {
  if (!all(c("mutType", "mutNum", "totalRootNum") %in% names(df))) {
    stop("TSV file is missing required columns (mutType, mutNum, totalRootNum)")
  }
  df <- df %>%
    mutate(
      mutNum = as.numeric(.data$mutNum),
      totalRootNum = as.numeric(.data$totalRootNum),
      observed = if_else(.data$totalRootNum != 0, .data$mutNum / .data$totalRootNum, 0)
    )
  mean_observed <- ifelse(nrow(df) > 0, mean(df$observed, na.rm = TRUE), 0)
  adjusted_mean <- max(mean_observed, EPS)
  ratios <- (df$observed + EPS) / adjusted_mean
  list(
    logratio = log2(ratios),
    observed = df$observed
  )
}

load_tsv_vectors <- function(file_path, reference_order = NULL) {
  df <- read_tsv(file_path, show_col_types = FALSE)
  if (is.null(reference_order)) {
    df_sorted <- df %>% arrange(.data$mutType)
    vectors <- compute_feature_vectors(df_sorted)
    order <- df_sorted$mutType
    return(list(logratio = vectors$logratio, observed = vectors$observed, order = order))
  }
  df_indexed <- df %>%
    filter(.data$mutType %in% reference_order) %>%
    arrange(match(.data$mutType, reference_order))
  if (!all(reference_order %in% df_indexed$mutType)) {
    stop(sprintf("TSV file %s does not contain expected mutType values", basename(file_path)))
  }
  vectors <- compute_feature_vectors(df_indexed)
  list(logratio = vectors$logratio, observed = vectors$observed, order = reference_order)
}

determine_clusters <- function(data, max_clusters, seed) {
  sample_count <- nrow(data)
  if (sample_count < 2) stop("At least two samples are required for clustering.")
  adjusted_max <- min(max_clusters, sample_count)
  if (adjusted_max < 2) adjusted_max <- 2

  set.seed(seed)
  inertia_values <- numeric()
  silhouette_scores <- numeric()

  for (k in 2:adjusted_max) {
    model <- kmeans(data, centers = k, nstart = 10)
    inertia_values <- c(inertia_values, model$tot.withinss)
    sil <- silhouette(model$cluster, dist(data))
    silhouette_scores <- c(silhouette_scores, mean(sil[, "sil_width"]))
  }

  best_k <- which.max(silhouette_scores) + 1
  list(best_k = best_k, inertia = inertia_values, silhouette = silhouette_scores)
}

plot_metrics <- function(output_dir, inertia, silhouette, suffix) {
  ks <- seq_along(inertia) + 1
  metrics_df <- tibble(
    k = ks,
    inertia = inertia,
    silhouette = silhouette
  )

  p1 <- ggplot(metrics_df, aes(x = .data$k, y = .data$inertia)) +
    geom_line(color = "blue") +
    geom_point(color = "blue") +
    labs(title = "Elbow Method", x = "Number of Clusters", y = "Inertia") +
    theme_minimal() +
    theme(
      panel.grid.major = element_line(color = "#d0d0d0", linewidth = 0.4),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black")
    )

  p2 <- ggplot(metrics_df, aes(x = .data$k, y = .data$silhouette)) +
    geom_line(color = "red") +
    geom_point(color = "red") +
    labs(title = "Silhouette Scores", x = "Number of Clusters", y = "Silhouette Score") +
    theme_minimal() +
    theme(
      panel.grid.major = element_line(color = "#d0d0d0", linewidth = 0.4),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black")
    )

  file_name <- if (suffix == "logratio") "clustering_metrics.png" else sprintf("clustering_metrics_%s.png", suffix)

  ggsave(
    file.path(output_dir, file_name),
    ggpubr::ggarrange(p1, p2, nrow = 1),
    dpi = 300,
    width = 12,
    height = 5
  )
}

plot_pca_scatter <- function(output_dir, pca_data, labels, genus_labels, cluster_labels, suffix) {
  df <- tibble(
    PC1 = pca_data[, 1],
    PC2 = pca_data[, 2],
    species = stringr::str_replace_all(labels, "_", " "),
    genus = genus_labels,
    cluster = factor(cluster_labels)
  )

  palette <- hue_pal()(length(unique(df$genus)))
  names(palette) <- unique(df$genus)

  subtitle_text <- if (suffix == "logratio") "Features: log2((#sbst/#ori)/mean of #sbst/#ori)" else "Features: #sbst/#ori"

  p <- ggplot(df, aes(x = .data$PC1, y = .data$PC2, color = .data$genus)) +
    geom_point(size = 3, alpha = 0.8) +
    ggrepel::geom_text_repel(
      aes(label = .data$species),
      size = 3,
      box.padding = 0.5,
      point.padding = 0.3,
      max.overlaps = Inf,
      segment.color = "gray50",
      segment.alpha = 0.7,
      fontface = "italic",
      force = 3,
      min.segment.length = 0
    ) +
    scale_color_manual(values = palette) +
    labs(
      title = sprintf("K-means Clustering (k=%d)", length(unique(cluster_labels))),
      subtitle = subtitle_text,
      x = "PC1",
      y = "PC2"
    ) +
    theme_minimal() +
    theme(
      panel.grid.major = element_line(color = "#d0d0d0", linewidth = 0.4),
      panel.grid.minor = element_blank(),
      axis.line = element_line(color = "black")
    )

  file_name <- if (suffix == "logratio") "pca_2d_clusters.png" else sprintf("pca_2d_clusters_%s.png", suffix)

  ggsave(file.path(output_dir, file_name), p, dpi = 300, width = 10, height = 8)
}

run_pca_workflow <- function(data_matrix, feature_suffix, output_dir, species_labels, genus_labels, max_clusters, random_state) {
  scaled_data <- scale(data_matrix)
  pca <- prcomp(scaled_data, center = FALSE, scale. = FALSE)
  pca_data <- pca$x[, 1:2, drop = FALSE]

  cluster_info <- determine_clusters(pca_data, max_clusters, random_state)
  plot_metrics(output_dir, cluster_info$inertia, cluster_info$silhouette, feature_suffix)

  set.seed(random_state)
  final_model <- kmeans(pca_data, centers = cluster_info$best_k, nstart = 10)
  cluster_labels <- final_model$cluster

  plot_pca_scatter(output_dir, pca_data, species_labels, genus_labels, cluster_labels, feature_suffix)

  list(
    cluster_labels = cluster_labels,
    pca_data = pca_data,
    best_k = cluster_info$best_k
  )
}

save_results <- function(output_dir, files, metadata_records, logratio_results, observed_results) {
  df <- tibble(
    file = basename(files),
    accession = map_chr(metadata_records, "accession"),
    short_name = map_chr(metadata_records, "short_name"),
    date = map_chr(metadata_records, "date"),
    genus = map_chr(metadata_records, "genus"),
    species = map_chr(metadata_records, "species"),
    cluster_logratio = as.integer(logratio_results$cluster_labels),
    cluster_observed = as.integer(observed_results$cluster_labels),
    PC1_logratio = logratio_results$pca_data[, 1],
    PC2_logratio = logratio_results$pca_data[, 2],
    PC1_observed = observed_results$pca_data[, 1],
    PC2_observed = observed_results$pca_data[, 2]
  )
  write_csv(df, file.path(output_dir, "clustering_results.csv"))
}

run <- function() {
  args <- parse_args()

  tsv_dir <- normalizePath(args$tsv_dir)
  if (!dir.exists(tsv_dir)) {
    stop(sprintf("TSV directory does not exist: %s", tsv_dir))
  }
  output_dir <- if (!is.null(args$output_dir)) normalizePath(args$output_dir) else file.path(tsv_dir, "pca_logratio_output")
  dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

  files <- collect_tsv_files(tsv_dir, args$tsv_pattern)

  logratio_vectors <- list()
  observed_vectors <- list()
  labels <- character()
  genus_labels <- character()
  metadata_records <- list()
  reference_order <- NULL

  for (file in files) {
    parts <- parse_filename_parts(file)
    metadata <- ensure_metadata(file, parts$accession)
    label_info <- extract_genus_label(metadata, parts$short_name, parts$accession)

    loaded <- load_tsv_vectors(file, reference_order)
    reference_order <- loaded$order

    logratio_vectors[[length(logratio_vectors) + 1]] <- loaded$logratio
    observed_vectors[[length(observed_vectors) + 1]] <- loaded$observed
    labels <- c(labels, label_info$species)
    genus_labels <- c(genus_labels, label_info$genus)
    metadata_records[[length(metadata_records) + 1]] <- list(
      accession = parts$accession,
      short_name = parts$short_name,
      date = parts$date,
      genus = label_info$genus,
      species = label_info$species
    )
  }

  logratio_matrix <- do.call(rbind, logratio_vectors)
  observed_matrix <- do.call(rbind, observed_vectors)

  logratio_results <- run_pca_workflow(
    logratio_matrix,
    "logratio",
    output_dir,
    labels,
    genus_labels,
    args$max_clusters,
    args$random_state
  )

  observed_results <- run_pca_workflow(
    observed_matrix,
    "observed",
    output_dir,
    labels,
    genus_labels,
    args$max_clusters,
    args$random_state
  )

  save_results(output_dir, files, metadata_records, logratio_results, observed_results)

  message("PCA clustering completed successfully.")
  message(sprintf("Processed files: %d", length(files)))
  message(sprintf("Optimal clusters (logratio k): %d", logratio_results$best_k))
  message(sprintf("Optimal clusters (observed k): %d", observed_results$best_k))
  message(sprintf("Results saved to: %s", output_dir))
}

tryCatch(
  run(),
  error = function(e) {
    message(sprintf("Error: %s", e$message))
    quit(status = 1)
  }
)
